File: stacks_and_queues/questions/question_3.7/problem_3.7.rb
require 'date'
class Node
  attr_accessor :value, :next

  def initialize(value)
    @value = value
    @next = nil
  end
end

class LinkedList
  attr_accessor :head

  def initialize(value=nil)
    @head = value ? Node.new(value) : nil
  end

  def peek
    @head ? @head.value : nil
  end

  def pop
    return nil unless @head
    return_value = @head.value
    @head = @head.next
    return_value
  end

  def push(value)
    new_node = Node.new(value)
    new_node.next = @head
    @head = new_node
  end
end

class Animal
  attr_accessor :name, :type, :date_added

  def initialize(name, type)
    @name = name
    @type = type
    @date_added = nil
  end
end



class AnimalShelter
  attr_accessor :dogs, :cats

  def initialize
    @dogs = LinkedList.new
    @cats = LinkedList.new
  end

  def enqueue(animal)
    animal.date_added = DateTime.now
    animal.type == 'dog' ? @dogs.push(animal) : @cats.push(animal)
  end

  def dequeue_dog
    @dogs.pop
  end

  def dequeue_cat
    @cats.pop
  end

  def dequeue_any
    dog = @dogs.peek
    cat = @cats.peek
    if dog && cat
      dog.date_added > cat.date_added ? @dogs.pop : @cats.pop
    else
      @dogs.pop || @cats.pop
    end
  end

end

a = LinkedList.new(5)
fido = Animal.new('fido', 'dog')
rex = Animal.new('rex', 'dog')
fluffy = Animal.new('fluffy', 'cat')
bandit = Animal.new('bandit', 'cat')

spca = AnimalShelter.new
spca.enqueue(fido)
spca.enqueue(fluffy)
spca.enqueue(bandit)
spca.enqueue(rex)
# p spca.dogs.peek
# p spca.dogs.peek.date_added > spca.cats.peek.date_added
p spca.dequeue_any
p spca.dequeue_any
p spca.dequeue_dog
p spca.dequeue_any
p spca.dequeue_any

File: stacks_and_queues/questions/question_3.7/problem_3.7.js
var Node = function(value){
  this.value = value
  this.next = null
}

var LinkedList = function(){
  this.head = null
}

LinkedList.prototype.peek = function(){
  return this.head ? this.head.value : null
}

LinkedList.prototype.pop = function(){
  if(!this.head){
    return null
  }
  popped_item = this.head.value
  this.head = this.head.next
  return popped_item
}

LinkedList.prototype.push = function(value){
  var newNode = new Node(value)
  newNode.next = this.head
  this.head = newNode
}

var Animal = function(name, type){
  this.name = name
  this.type = type
  this.timeAdded = null
}

var AnimalShelter = function(){
  this.dogs = new LinkedList()
  this.cats = new LinkedList()
}

AnimalShelter.prototype.enqueue = function(animal){
  // -------------------------------------------------------
  // because i am using the Date.now() function to timestamp
  // the addition of the new animal, i need to ensure that
  // the timestamps are all unique - this introduces a 1ms
  // pause between rapid enquees
  var pause = Date.now()
  while(pause == Date.now()){}
  // -------------------------------------------------------
  animal.timeAdded = Date.now()
  if(animal.type == 'dog'){
    this.dogs.push(animal)
  } else{
    this.cats.push(animal)
  }
}

AnimalShelter.prototype.dequeueDog = function(){
  return this.dogs.peek() ? this.dogs.pop() : null
}

AnimalShelter.prototype.dequeueCat = function(){
  return this.cats.peek() ? this.cats.pop() : null
}

AnimalShelter.prototype.dequeueAny = function(){
  var dog = this.dogs.peek()
  var cat = this.cats.peek()
  if(dog && cat){
    return dog.timeAdded > cat.timeAdded ? this.dogs.pop() : this.cats.pop()
  }
  return this.dogs.pop() || this.cats.pop()
}


var fido = new Animal('fido', 'dog')
var rex = new Animal('rex', 'dog')
var bandit = new Animal('bandit', 'cat')
var fluffy = new Animal('fluffy', 'cat')

aS = new AnimalShelter()
aS.enqueue(fido)
aS.enqueue(bandit)
aS.enqueue(fluffy)
aS.enqueue(rex)
// console.log(aS.dequeueAny())
// console.log(aS.dequeueAny())
// console.log(aS.dequeueAny())

console.log(aS.dogs.peek().timeAdded)
console.log(aS.cats.peek().timeAdded)
console.log(aS.dogs.peek().timeAdded > aS.cats.peek().timeAdded)
console.log(aS.dequeueAny())
console.log(aS.dequeueAny())
console.log(aS.dequeueAny())
console.log(aS.dequeueAny())
console.log(aS.dequeueAny())
